src/Common/getNumberOfPhysicalCPUCores.cpp

```diff
@@ -31,10 +34,89 @@ int32_t readFrom(const char * filename, int default_value)
 uint32_t getCGroupLimitedCPUCores(unsigned default_cpu_count)
 {
     uint32_t quota_count = default_cpu_count;
+    std::filesystem::path prefix = "/sys/fs/cgroup";
+    /// cgroupsv2
+    std::ifstream contr_file(prefix / "cgroup.controllers");
+    if (contr_file.is_open())
+    {
+        /// First, we identify the cgroup the process belongs
+        std::ifstream cgroup_name_file("/proc/self/cgroup");
+        if (!cgroup_name_file.is_open())
+            return default_cpu_count;
+
+        // cgroup_name_file always starts with '0::' for v2
+        cgroup_name_file.ignore(3);
+        std::string cgroup_name;
+        cgroup_name_file >> cgroup_name;
+        // cgroup_name always starts with /
+        cgroup_name = cgroup_name.substr(1);
+
+        std::filesystem::path current_cgroup;
+        if (cgroup_name.empty())
+            current_cgroup = prefix;
+        else
+            current_cgroup = prefix / cgroup_name;
+
+        // Looking for cpu.max in directories from the current cgroup to the top level
+        // It does not stop on the first time since the child could have a greater value than parent
+        while (current_cgroup != prefix.parent_path())
+        {
+            std::ifstream cpu_max_file(current_cgroup / "cpu.max");
+            current_cgroup = current_cgroup.parent_path();
+            if (cpu_max_file.is_open())
+            {
+                std::string cpu_limit_str;
+                float cpu_period;
+                cpu_max_file >> cpu_limit_str >> cpu_period;
+                if (cpu_limit_str != "max" && cpu_period != 0)
+                {
+                    float cpu_limit = std::stof(cpu_limit_str);
+                    quota_count = std::min(static_cast<uint32_t>(ceil(cpu_limit / cpu_period)), quota_count);
+                }
+            }
+        }
+        current_cgroup = prefix / cgroup_name;
+        // Looking for cpuset.cpus.effective in directories from the current cgroup to the top level
+        while (current_cgroup != prefix.parent_path())
```

Comment:
Do we really need this cycle if we already use `.effective`?

Reply:
yes, it's necessary

```
start_dir=/sys/fs/cgroup/user.slice/user-1000.slice/user@1000.service/app.slice/app-org.wezfurlong.wezterm-084ffa47df3f4bb692567dedc8b856e4.scope
while :; do ls ${start_dir}/cpuset.cpus.effective && break || start_dir=$(readlink -f "${start_dir}"/..)
done
ls: cannot access '/sys/fs/cgroup/user.slice/user-1000.slice/user@1000.service/app.slice/app-org.wezfurlong.wezterm-084ffa47df3f4bb692567dedc8b856e4.scope/cpuset.cpus.effective': No such file or directory
ls: cannot access '/sys/fs/cgroup/user.slice/user-1000.slice/user@1000.service/app.slice/cpuset.cpus.effective': No such file or directory
ls: cannot access '/sys/fs/cgroup/user.slice/user-1000.slice/user@1000.service/cpuset.cpus.effective': No such file or directory
ls: cannot access '/sys/fs/cgroup/user.slice/user-1000.slice/cpuset.cpus.effective': No such file or directory
/sys/fs/cgroup/user.slice/cpuset.cpus.effective
```

There are no guarantees that the file exists in the `/sys/fs/cgroup/$(/proc/self/cgroup)` directory

Reply:
Ah, we `break` after the first match. I see...